### DISCUSSION TITLE
Separate For variables for DatasourceNoData and DatasourceError alerts

### DISCUSSION THREAD

**Discussion by grobinson-grafana** [grafana/grafana#44725] [Alerting]
Grafana supports two different kind of alerts for when we are either missing data or cannot evaluate an alert rule due to some error. These are `DatasourceNoData` and `DatasourceError`.

However, while all other alerts wait until `For` seconds have elapsed before transitioning from `pending` to `firing`, `DatasourceNoData` and `DatasourceError` will alert on the first occurrence of missing data or an error and it is not possible to set a `For` variable for either of these alerts. This means that both Grafana Cloud and open source users can receive `DatasourceError` alerts for short-lived ephemeral errors.

While we could use the existing `For` variable for `DatasourceNoData` and `DatasourceError` alerts, it can be problematic for alert rules that have long evaluation intervals, such as 15m or 1h. We are thinking of creating separate `For` variables for both `DatasourceNoData` and `DatasourceError`, which like the existing `For` variable, is configurable for each alert rule rather than all alert rules.

**Comment by fleetwoodstack** [comment]
Yes yes yes please implement this. We have to switch off No Data errors because of this missing feature. I know it must be down to some sort of flaky connection that the data source has but having a 'For' variable for data source errors/no data alerts would be so helpful!

**Comment by itsham-sajid** [comment]
Second the above, it's a useful feature for addressing the issue of short-lived errors and missing data alerts.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
